chainstate/src/detail/spend_cache/mod.rs

```diff
@@ -205,15 +209,105 @@ impl<'a, S: BlockchainStorageRead> CachedInputs<'a, S> {
         outputs: &[TxOutput],
         output_index: usize,
         spender_id: Spender,
-    ) -> Result<Amount, StateUpdateError> {
+    ) -> Result<&OutputValue, StateUpdateError> {
         let output = outputs.get(output_index).ok_or(StateUpdateError::OutputIndexOutOfRange {
             tx_id: Some(spender_id),
             source_output_index: output_index,
         })?;
         Ok(output.value())
     }
 
-    fn calculate_total_inputs(&self, inputs: &[TxInput]) -> Result<Amount, StateUpdateError> {
+    // Get TokenId and Amount in input
+    fn filter_assets_amount(
+        &self,
+        prev_tx: &Transaction,
+        output: &common::chain::TxOutput,
+    ) -> Option<(TokenId, Amount)> {
+        match output.value() {
+            OutputValue::Coin(_) => None,
+            OutputValue::Asset(asset) => Some(match asset {
+                AssetData::TokenTransferV1 { token_id, amount } => (*token_id, *amount),
+                AssetData::TokenIssuanceV1 {
+                    token_ticker: _,
+                    amount_to_issue,
+                    number_of_decimals: _,
+                    metadata_uri: _,
+                } => {
+                    let token_id = token_id(prev_tx)?;
+                    (token_id, *amount_to_issue)
+                }
+                AssetData::TokenBurnV1 {
+                    token_id: _,
+                    amount_to_burn: _,
+                } => {
+                    /* Token have burned and can't be transfered */
+                    return None;
+                }
+            }),
+        }
+    }
+
+    pub fn calculate_assets_total_inputs(
+        &self,
+        inputs: &[TxInput],
+    ) -> Result<BTreeMap<TokenId, Amount>, StateUpdateError> {
+        let mut total_assets: BTreeMap<TokenId, Amount> = BTreeMap::new();
+        for (_input_idx, input) in inputs.iter().enumerate() {
+            let outpoint = input.outpoint();
+            let tx_index = match self.inputs.get(&outpoint.tx_id()) {
```

Comment:
Since this call is done multiple times around, abstract it away with a function call and use `?` to return an error, if any.

Reply:
Done